### PR TITLE
allow to specify expected stdout in test

### LINKED ISSUE #101: allow to specify the expected stdout of a tested script

### PR DESCRIPTION
**Pull request by hallettj**
fixes #101 

This PR also changes the `test_run` family of test helpers to take a value of a new type, `Expect`, which provides more flexibility for specifying expected results.

### CODE REVIEW

**Comment by soenkehahn** [comment]
Could you squash all the changes related to `Expect` into one commit? I think that would make it really easy to review this PR.

**Comment by hallettj** [comment]
@soenkehahn Sure thing - I squashed and reordered some commits.

File: tests/utils.rs [hunk 1]
```diff
@@ -46,15 +81,22 @@ pub fn test_run_with_context(
     context: &Context,
```
Comment (soenkehahn) [inline tests/utils.rs]: Do we still need `test_run_with_context` to be public? Or can we replace that with `test_run` everywhere, new that matching results is smarter?

File: tests/utils.rs [hunk 2]
```diff
@@ -14,12 +14,47 @@ use std::path::PathBuf;
...
 use yaml_rust::YamlLoader;
 
-fn compare_results(result: (ExitCode, String), expected: Result<(), &str>) {
-    let expected_output = match expected {
-        Err(expected_output) => (ExitCode(1), expected_output.to_string()),
-        Ok(()) => (ExitCode(0), "All tests passed.\n".to_string()),
-    };
-    assert_eq!(result, expected_output);
+#[derive(Debug, PartialEq)]
+pub struct Expect {
```
Comment (matthandlersux) [inline tests/utils.rs]: "Expect" feels a bit generic for what this does. the verbose description i can think of is "ScriptRunEndState", but that's probably too wordy. any ideas?

File: tests/utils.rs [hunk 3]
```diff
@@ -14,12 +14,47 @@ use std::path::PathBuf;
...
+
+    pub fn err(expected_output: &str) -> Self {
+        Expect {
+            expected_exit_code: ExitCode(1),
+            expected_stdout: expected_output.to_string(),
+            expected_stderr: "".to_string(),
+        }
+    }
+
+    pub fn stdout(self, expected_output: &str) -> Self {
```
Comment (matthandlersux) [inline tests/utils.rs]: since the usage is `Expect::ok().stdout(...)` my vote would be to name this in a way that suggests you're adding to the structure rather than performing some action, something like `Expect::ok().with_stdout(...)`.

File: src/test_checker/mod.rs [hunk 4]
```diff
@@ -70,6 +70,28 @@ impl TestChecker {
...
+    fn check_expected_output_stream(&mut self, redirect: &Redirect, expected: Vec<u8>) -> R<()> {
+        match redirect.captured()? {
+            None => panic!(
+                "scriptkeeper bug: {} expected, but not captured",
+                redirect.stream_type
+            ),
+            Some(captured) => {
+                if captured != expected {
+                    self.register_error(format!(
+                        "  expected output to {}: {:?}\
```
Comment (matthandlersux) [inline src/test_checker/mod.rs]: i realize this was moved from below, but the `trim_margin` function is nice if we can use it here, i think it makes things much more readable

File: tests/utils.rs [hunk 5]
```diff
@@ -14,12 +14,47 @@ use std::path::PathBuf;
...
+        Expect {
+            expected_stderr: expected_output.to_string(),
+            ..self
+        }
+    }
+}
+
+fn compare_results(actual: Expect, expected: Expect) {
+    assert_eq!(actual, expected);
 }
```
[the remark this answers is not in the document]
Comment (hallettj) [reply tests/utils.rs]: Good point, inlined.

File: tests/utils.rs (hunk 1 above)
Comment (hallettj) [reply tests/utils.rs]: Yes, and we can also inline `test_run_with_context`.

File: src/tracer/stdio_redirecting.rs [hunk 6]
```diff
@@ -47,44 +45,41 @@ impl Redirector {
...
     write_end: RawFd,
     captured: Option<Arc<Mutex<Cursor<Vec<u8>>>>>,
 }
 
 impl Redirect {
-    fn new(context: &Context, stream_type: StreamType) -> R<Redirect> {
-        Redirect::new_internal(context, stream_type, None)
-    }
-
-    fn new_capturing(context: &Context, stream_type: StreamType) -> R<Redirect> {
```
[the remark this answers is not in the document]
Comment (hallettj) [reply src/tracer/stdio_redirecting.rs]: Ok, I split out those two constructor variants.

File: tests/stdio.rs [hunk 7]
```diff
@@ -49,12 +39,94 @@ fn relays_stderr_from_the_tested_script_to_the_user() -> R<()> {
...
+                    |  expected output to stdout: "foo\n"
+                    |  received output to stdout: ""
+                "#,
+            )?),
+        )?;
+        Ok(())
+    }
+
+    #[test]
+    fn when_not_specified_but_scripts_writes_to_stdout() -> R<()> {
```
[the remark this answers is not in the document]
Comment (hallettj) [reply tests/stdio.rs]: Done

File: tests/stdio.rs [hunk 8]
```diff
@@ -119,4 +192,20 @@ mod expected_stderr {
         )?;
         Ok(())
     }
+
+    #[test]
+    fn when_not_specified_but_scripts_writes_to_stderr() -> R<()> {
```
[the remark this answers is not in the document]
Comment (hallettj) [reply tests/stdio.rs]: I reworded this test name and the name of the corresponding stdout test a little bit. The detail we are testing is that scriptkeeper tests pass when there is output if no expected output has been specified.

File: tests/utils.rs (hunk 2 above)
Comment (hallettj) [reply tests/utils.rs]: My preference is to put descriptiveness in the method names, as opposed to the name of the type itself. To that end I changed the names of the `Expect` methods:

- `ok` → `tests_pass`
- `err` → `error_message`
- `stdout` → `with_stdout` (per your suggestion)
- `stderr` → `with_stderr`

Although this type is specialized for our domain, I think it would be nice for it to read like a sentence as expectations in some test frameworks do. For example, `Expect::error_message("failure").with_stdout("test")`

File: tests/utils.rs (hunk 3 above)
Comment (hallettj) [reply tests/utils.rs]: Good idea! I have made that change.

File: src/test_checker/mod.rs (hunk 4 above)
Comment (hallettj) [reply src/test_checker/mod.rs]: Sounds good. I moved the trim-margin dependency from the build dependencies list to general dependencies to make this work.

File: tests/run.rs [hunk 9]
```diff
@@ -198,7 +198,7 @@ fn failing() -> R<()> {
             |steps:
             |  - cp
         ",
-        Err(&trim_margin(
+        Expect::error_message(&trim_margin(
```
Comment (soenkehahn) [inline tests/run.rs]: Do we want to move the call to `trim_margin` into `Expect::error_message`? Do we ever use `error_message` without it?

File: tests/utils.rs [hunk 10]
```diff
@@ -42,22 +73,21 @@ pub fn test_run_with_tempfile(
     Ok((exitcode, context.get_captured_stdout()))
 }
```
Comment (soenkehahn) [inline tests/utils.rs]: Could `test_run_with_tempfile` now be changed to not return `(ExitCode, String)` and instead take an argument of type `Expect`?

File: src/test_checker/mod.rs [hunk 11]
```diff
@@ -70,6 +71,35 @@ impl TestChecker {
...
+                    );
+                }
+            }
+        }
+        Ok(())
+    }
+
     fn register_step_error(&mut self, expected: &str, received: &str) {
         self.register_error(format!(
             "  expected: {}\n  received: {}\n",
```
[the remark this answers is not in the document]
Comment (hallettj) [reply src/test_checker/mod.rs]: Sure thing, I've made that change.

File: tests/run.rs (hunk 9 above)
Comment (hallettj) [reply tests/run.rs]: You're right, we use `trim_margin` in every case where we use `Expect::error_message` I made this change; and now `Expect::error_message` returns `R<Expect>` instead of `Expect`.

File: tests/utils.rs (hunk 10 above)
Comment (hallettj) [reply tests/utils.rs]: As we discussed passing an `Expect` to `test_run_with_tempfile` does not work out. But I simplified `test_run_with_tempfile` to return just an exit code.

**Comment by hallettj** [comment]
@soenkehahn How do things look after my latest set of changes?

**Comment by matthandlersux** [comment]
@soenkehahn i believe this PR is just waiting on your final review

**Comment by soenkehahn** [comment]
Yeah, sorry. I think finishing up the first binary release has a higher priority right now. I hope I'll look at this soon.

**Comment by soenkehahn** [comment]
@hallettj: I took the liberty to merge more commits and rebase on current master. Mostly I combined later refactorings into commits that introduced the code change in the first place. I think I'd like to have the introduction of `Expect` and the support for expected `stdout` as separate commits on master, so I didn't want to squash this PR. But then I also didn't want to have all these small refactoring commits you had before.

It's also much easier to review now. And LGTM. Should I merge or do you still have any comments/changes?

**Comment by hallettj** [comment]
Sounds good. If it looks good to you then please feel free to merge.

**Comment by soenkehahn** [comment]
@hallettj: Thanks a lot for the PR. :)
